core/primitives/src/types.rs

```diff
@@ -152,6 +154,28 @@ pub enum StateChangeCause {
     /// State change that happens when we update validator accounts. Not associated with with any
     /// specific transaction or receipt.
     ValidatorAccountsUpdate,
+    /// State change that is happens due to migration that happens in first block of an epoch
+    /// after protocol upgrade
+    Migration { migration_id: MigrationId },
+}
+
+#[derive(Serialize, Deserialize, Clone, BorshSerialize, BorshDeserialize)]
+pub enum MigrationId {
+    StorageUsageFix,
+}
+
+impl fmt::Display for MigrationId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::StorageUsageFix => write!(f, "Storage usage fix"),
+        }
+    }
+}
+
+impl fmt::Debug for MigrationId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        (self as &dyn fmt::Display).fmt(f)
```

Comment:
`wirite!(f, "{}", self)` or `fmt::Display::fmt(self, f)` would be a more direct way to write this.

Reply:
I removed Display anyway as it is not really needed, only Debug

Reply:
Apparently it is needed after all, done